src/keystore.c

```diff
@@ -106,39 +196,73 @@ void DPS_DestroyMemoryKeyStore(DPS_MemoryKeyStore* mks)
         DPS_MemoryKeyStoreEntry* entry = mks->entries + i;
         free(entry->key);
     }
-    free(mks->entries);
-    free(mks->networkKey);
+    if (mks->entries) {
+        free(mks->entries);
+    }
+    if (mks->networkId) {
+        free(mks->networkId);
+    }
+    if (mks->networkKey) {
+        free(mks->networkKey);
+    }
+    if (mks->ca) {
+        free(mks->ca);
+    }
+    if (mks->cert) {
+        free(mks->cert);
+    }
+    if (mks->privateKey) {
+        free(mks->privateKey);
+    }
+    if (mks->password) {
+        free(mks->password);
+    }
     free(mks);
 }
 
-DPS_Status DPS_SetNetworkKey(DPS_MemoryKeyStore* mks, uint8_t* key, size_t keyLen)
+DPS_Status DPS_SetNetworkKey(DPS_MemoryKeyStore* mks, const uint8_t* id, size_t idLen, const uint8_t* key, size_t keyLen)
 {
     DPS_DBGTRACE();
 
-    if ((!key && keyLen > 0) || (key && keyLen == 0)) {
+    if (!id || (idLen == 0) || (!key && keyLen > 0) || (key && keyLen == 0)) {
         return DPS_ERR_INVALID;
     }
 
+    uint8_t* networkId = networkId = calloc(1, idLen);
+    if (!networkId) {
+        return DPS_ERR_RESOURCES;
+    }
+    memcpy(networkId, id, idLen);
+
     uint8_t* networkKey = NULL;
     if (key) {
         networkKey = calloc(1, keyLen);
         if (!networkKey) {
+            free(mks->networkId);
             return DPS_ERR_RESOURCES;
         }
         memcpy(networkKey, key, keyLen);
     }
 
-    free(mks->networkKey);
+    if (mks->networkId) {
+        free(mks->networkId);
+    }
+    mks->networkId = networkId;
+    mks->networkIdLen = idLen;
+    if (mks->networkKey) {
+        free(mks->networkKey);
+    }
     mks->networkKey = networkKey;
     mks->networkKeyLen = keyLen;
+
     return DPS_OK;
 }
 
-DPS_Status DPS_SetContentKey(DPS_MemoryKeyStore* mks, const DPS_UUID* kid, uint8_t* key, size_t keyLen)
+DPS_Status DPS_SetContentKey(DPS_MemoryKeyStore* mks, const DPS_UUID* kid, const uint8_t* key, size_t keyLen)
 {
     DPS_DBGTRACE();
 
-    // Replace key if kid already exists.
+    /* Replace key if kid already exists. */
     for (size_t i = 0; i < mks->entriesCount; i++) {
         DPS_MemoryKeyStoreEntry* entry = mks->entries + i;
```

Comment:
could this be accessing one past the end of mks->entries?

Reply:
i < mks->entriesCount so I think this is safe